library/Icingadb/Common/Macros.php

```diff
@@ -0,0 +1,103 @@
+<?php
+
+/* Icinga DB Web | (c) 2021 Icinga GmbH | GPLv2 */
+
+namespace Icinga\Module\Icingadb\Common;
+
+use Icinga\Application\Logger;
+use Icinga\Module\Icingadb\Model\Host;
+use Icinga\Module\Icingadb\Model\Service;
+use ipl\Orm\Query;
+use ipl\Orm\ResultSet;
+
+trait Macros
+{
+    /**
+     * Get the given string with macros being resolved
+     *
+     * @param string       $input  The string in which to look for macros
+     * @param Host|Service $object The host or service used to resolve the macros
+     *
+     * @return string
+     */
+    public function expandMacros(string $input, $object): string
+    {
+        if (preg_match_all('@\$([^\$\s]+)\$@', $input, $matches)) {
+            foreach ($matches[1] as $key => $value) {
+                $newValue = $this->resolveMacro($value, $object);
+                if ($newValue !== $value) {
+                    $input = str_replace($matches[0][$key], $newValue, $input);
+                }
+            }
+        }
+
+        return $input;
+    }
+
+    /**
+     * Resolve a macro based on the given object
+     *
+     * @param string       $macro  The macro to resolve
+     * @param Host|Service $object The host or service used to resolve the macros
+     *
+     * @return string
+     */
+    public function resolveMacro(string $macro, $object): string
+    {
+        if ($object instanceof Host) {
+            $objectType = 'host';
+        } else {
+            $objectType = 'service';
+        }
+
+        $macroType = null;
+        $isCustomVar = false;
+        if (preg_match('/^((host|service)[.])?vars[.](.+)/i', $macro, $matches)) {
```

Comment:
Why `[.]` and not `\.`?

And please don't make it case-insensitive. We process `host`, `service` or `vars` case-sensitive everywhere else. We shouldn't accept e.g. `HOST` here either.